server/src/main/java/org/opensearch/common/lucene/index/OpenSearchDirectoryReader.java

```diff
@@ -56,13 +72,52 @@ public final class OpenSearchDirectoryReader extends FilterDirectoryReader {
     private final FilterDirectoryReader.SubReaderWrapper wrapper;
 
     private final DelegatingCacheHelper delegatingCacheHelper;
+    // Per criteria segments to avoid runtime penalty during search
+    private Map<String, OpenSearchDirectoryReader> criteriaBasedReaders;
+    private OpenSearchDirectoryReader emptyCriteriaBasedReader;
+
+    /**
+     * {@link org.opensearch.indices.IndicesRequestCache} depends on cache key generated in OpenSearchDirectoryReader
+     * This cache is used to be able to keep the cacheKey of OpensearchDirectoryReader consistent in cases of multiple criteria
+     */
+    private Cache<Set<String>, OpenSearchDirectoryReader> criteriaCache;
```

Comment:
Based on the extractCriteria implementation, the key `Set<String>` will be ordered based on how they appear in the query if there are multiple criterias ( BooleanQuery )

I think we need to preserve the order for the cache lookup to work if query is same but order of criteria fields are different.